gateway/middleware_ratelimit_test.go

```diff
@@ -0,0 +1,250 @@
+			for {
+				max := maxConcurrent.Load()
+				if cur <= max || maxConcurrent.CompareAndSwap(max, cur) {
+					break
+				}
+			}
+
+			time.Sleep(50 * time.Millisecond)
+			w.Write([]byte("success"))
+		}), limit, nil, newTestMetrics())
+
+		// Send more requests than the limit
+		var wg sync.WaitGroup
+		results := make([]int, 5)
+		for i := 0; i < 5; i++ {
+			wg.Add(1)
+			go func(idx int) {
+				defer wg.Done()
+				req := httptest.NewRequest(http.MethodGet, "/", nil)
+				rec := httptest.NewRecorder()
+				handler.ServeHTTP(rec, req)
+				results[idx] = rec.Code
+			}(i)
+		}
+		wg.Wait()
+
+		// Check that max concurrent never exceeded limit
+		if maxConcurrent.Load() > int32(limit) {
+			t.Errorf("max concurrent requests %d exceeded limit %d", maxConcurrent.Load(), limit)
+		}
+
+		// Count successful and rejected requests
+		successCount := 0
+		rejectCount := 0
+		for _, code := range results {
+			if code == http.StatusOK {
+				successCount++
+			} else if code == http.StatusTooManyRequests {
+				rejectCount++
+			}
+		}
+
+		if successCount == 0 {
+			t.Error("expected at least some successful requests")
+		}
+		if rejectCount == 0 {
+			t.Error("expected at least some rejected requests when limit exceeded")
+		}
+	})
+
+	t.Run("returns 429 with Retry-After and Cache-Control headers", func(t *testing.T) {
+		handler := withConcurrentRequestLimiter(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			time.Sleep(100 * time.Millisecond)
+			w.Write([]byte("success"))
+		}), 1, nil, newTestMetrics())
+
+		// First request should succeed
+		req1 := httptest.NewRequest(http.MethodGet, "/", nil)
+		rec1 := httptest.NewRecorder()
+		go func() {
+			handler.ServeHTTP(rec1, req1)
+		}()
+
+		// Give first request time to start
+		time.Sleep(10 * time.Millisecond)
+
+		// Second request should be rejected
+		req2 := httptest.NewRequest(http.MethodGet, "/", nil)
+		rec2 := httptest.NewRecorder()
+		handler.ServeHTTP(rec2, req2)
+
+		if rec2.Code != http.StatusTooManyRequests {
+			t.Errorf("expected status 429, got %d", rec2.Code)
+		}
+
+		retryAfter := rec2.Header().Get("Retry-After")
+		if retryAfter == "" {
+			t.Error("expected Retry-After header to be set")
+		}
+
+		retrySeconds, err := strconv.Atoi(retryAfter)
+		if err != nil {
+			t.Errorf("Retry-After should be a number, got %s", retryAfter)
+		}
+
+		// Retry should be 60 seconds (industry standard minimum)
+		if retrySeconds != 60 {
+			t.Errorf("expected Retry-After to be 60, got %d", retrySeconds)
+		}
+
+		// Check Cache-Control header to prevent caching
+		cacheControl := rec2.Header().Get("Cache-Control")
+		if cacheControl != "no-store" {
+			t.Errorf("expected Cache-Control: no-store, got %s", cacheControl)
+		}
+	})
+
+	t.Run("Retry-After is static", func(t *testing.T) {
+		handler := withConcurrentRequestLimiter(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			// Hold the request indefinitely
+			select {}
```

Comment:
Release it after the test is run, just so as not to hang goroutines during test.
```suggestion
		done := make(chan strict{})
		defer close(done)
		handler := withConcurrentRequestLimiter(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			// Hold the request indefinitely
			<-done
```